src/compiler/binder.ts

```diff
@@ -567,174 +570,501 @@ namespace ts {
+                        return true;
+                    }
+                    return false;
+                case SyntaxKind.InstanceOfKeyword:
+                    return isNarrowingExpression(expr.left);
+            }
+            return false;
         }
 
-        function bindForInOrForOfStatement(n: ForInStatement | ForOfStatement): void {
-            const preStatementState = currentReachabilityState;
-            const postStatementLabel = pushImplicitLabel();
-
-            // bind expressions (don't affect reachability)
-            bind(n.initializer);
-            bind(n.expression);
+        function createFlowLabel(): FlowLabel {
+            return {
+                kind: FlowKind.Label,
+                antecedents: undefined
+            };
+        }
 
-            bind(n.statement);
-            popImplicitLabel(postStatementLabel, preStatementState);
+        function addAntecedent(label: FlowLabel, antecedent: FlowNode): void {
+            if (antecedent.kind !== FlowKind.Unreachable && !contains(label.antecedents, antecedent)) {
+                (label.antecedents || (label.antecedents = [])).push(antecedent);
+            }
         }
 
-        function bindIfStatement(n: IfStatement): void {
-            // denotes reachability state when entering 'thenStatement' part of the if statement:
-            // i.e. if condition is false then thenStatement is unreachable
-            const ifTrueState = n.expression.kind === SyntaxKind.FalseKeyword ? Reachability.Unreachable : currentReachabilityState;
-            // denotes reachability state when entering 'elseStatement':
-            // i.e. if condition is true then elseStatement is unreachable
-            const ifFalseState = n.expression.kind === SyntaxKind.TrueKeyword ? Reachability.Unreachable : currentReachabilityState;
+        function createFlowCondition(antecedent: FlowNode, expression: Expression, assumeTrue: boolean): FlowNode {
+            if (antecedent.kind === FlowKind.Unreachable) {
+                return antecedent;
+            }
+            if (!expression) {
+                return assumeTrue ? antecedent : unreachableFlow;
+            }
+            if (expression.kind === SyntaxKind.TrueKeyword && !assumeTrue || expression.kind === SyntaxKind.FalseKeyword && assumeTrue) {
+                return unreachableFlow;
+            }
+            if (!isNarrowingExpression(expression)) {
+                return antecedent;
+            }
+            return <FlowCondition>{
+                kind: FlowKind.Condition,
+                antecedent,
+                expression,
+                assumeTrue
+            };
+        }
 
-            currentReachabilityState = ifTrueState;
+        function createFlowAssignment(antecedent: FlowNode, node: Expression | VariableDeclaration | BindingElement): FlowNode {
+            return <FlowAssignment>{

```

Comment:
Could you simply use these types as return types instead of using a type assertion?